modules/schema-util/src/main/scala/edu/gemini/schema/util/SchemaStitcher.scala

```diff
@@ -0,0 +1,177 @@
+      .map(x => collapseToList(AllElements, x))
+      .map(x => merge(x, List.empty))
+      .map(_.map(_.asString).mkString("\n"))
+      .map(Schema.apply(_))
+
+    def dependenciesTree(pathToRoot: List[Path], schemaName: Path): F[DependencyNode] =
+      if (pathToRoot.contains(schemaName)) {
+        Sync[F].raiseError(
+          new Exception(s"Found circular reference when resolving schema $schemaName")
+        )
+      } else {
+        resolver.resolve(schemaName).map(_.getLines.toList).use { ll =>
+          ll.map(importLineParser.parse)
+            .collect { case Right((_, (els, path))) =>
+              dependenciesTree(pathToRoot :+ schemaName, path).map((els, _))
+            }
+            .sequence
+            .map(u => DependencyNode(schemaName, ll, u))
+        }
+      }
+
+    def collapseToList(els: Elements, node: DependencyNode): List[SchemaNode] =
+      node.dependencies.flatMap { case (s, n) => collapseToList(s, n) } :+ SchemaNode(
+        node.v,
+        node.src.mkString("\n"),
+        els
+      )
+
+    def merge(l: List[SchemaNode], acc: List[SchemaNode]): List[SchemaNode] =
+      l match {
+        case Nil      => acc
+        case x :: Nil => acc :+ x
+        case x :: xx  =>
+          val (m, r) = xx.partition(_.name === x.name)
+          val s      = m.fold(x)((a, b) => SchemaNode(a.name, a.src, a.elements.union(b.elements)))
+          merge(r, acc :+ s)
+      }
+  }
+
+  sealed trait Elements
+  case object AllElements                                      extends Elements
+  final case class ElementList(l: NonEmptySet[NonEmptyString]) extends Elements
+
+  extension (el: Elements) {
+    def union(other: Elements): Elements = (el, other) match {
+      case (ElementList(l1), ElementList(l2)) => ElementList(l1 ++ l2)
+      case _                                  => AllElements
+    }
+  }
+
+  case class DependencyNode(
+    v:            Path,
+    src:          List[String],
+    dependencies: List[(Elements, DependencyNode)]
+  )
+
+  case class SchemaNode(name: Path, src: String, elements: Elements) {
+    def asString: String = elements match {
+      case AllElements    => src
+      case ElementList(l) => asString(l)
+    }
+
+    def asString(l: NonEmptySet[NonEmptyString]): String =
+      Schema(src) match {
+        case Ior.Right(b)   =>
+          resolveTypes(b.types,
+                       l.toList.map(x => b.types.find(_.name === x.toString)).flattenOption,
+                       List.empty
+          )
+            .map(SchemaRenderer.renderTypeDefn)
+            .mkString("\n")
+        case Ior.Both(_, b) =>
+          resolveTypes(b.types,
+                       l.toList.map(x => b.types.find(_.name === x.toString)).flattenOption,
+                       List.empty
+          )
+            .map(SchemaRenderer.renderTypeDefn)
+            .mkString("\n")
+        case _              => ""
+      }
+
+    def resolveTypes(
```

Comment:
resolveTypes uses Grackle data types. I worry that, 1) I don't really understand them, and 2) they may change in the future.